core/src/main/java/org/apache/iceberg/avro/ValueWriters.java

```diff
@@ -369,17 +393,42 @@ private OptionWriter(int nullIndex, ValueWriter<T> valueWriter) {
         throw new IllegalArgumentException("Invalid option index: " + nullIndex);
       }
       this.valueWriter = valueWriter;
+      this.type = type;
+      this.nullValueCount = 0;
     }
 
     @Override
     public void write(T option, Encoder encoder) throws IOException {
       if (option == null) {
         encoder.writeIndex(nullIndex);
+        nullValueCount++;
       } else {
         encoder.writeIndex(valueIndex);
         valueWriter.write(option, encoder);
       }
     }
+
+    @Override
+    public Stream<FieldMetrics> metrics() {
+      if (AvroSchemaUtil.isMetricSupportedType(type)) {
+        return mergeNullCountIntoMetric();
+      } else {
+        return valueWriter.metrics();
+      }
+    }
+
+    private Stream<FieldMetrics> mergeNullCountIntoMetric() {
+      List<FieldMetrics> fieldMetricsFromWriter = valueWriter.metrics().collect(Collectors.toList());
+      Preconditions.checkState(fieldMetricsFromWriter.size() == 1,
+          "Optional field for type % shouldn't vend more than one field metrics", type);
```

Comment:
Typo: should be `%s` instead of `%`.

Reply:
Also, the use of "vend" is unusual. What about "Optional %s writer should not produce metrics for more than one field"

Reply:
Okay, thinking about this a bit more, I don't think this assumption holds.

There are lots of times when a structure in the middle of a record might be optional. An optional map, for example, would have both key and value metrics. I think that we should only merge null in if we know that this is wrapping a leaf writer, not a map, list, or struct writer.

Reply:
I think this is currently handled by its caller, that only calls this if `AvroSchemaUtil. isMetricSupportedType` is true, and `AvroSchemaUtil. isMetricSupportedType` only returns true for primitive types?

Reply:
Yes, you're right that it is handled there. I missed that originally. I commented on how to avoid passing the type in, but I agree that you do handle it.